internal/app/init/internal/reg/reg.go

```diff
@@ -73,7 +73,19 @@ func (r *Registrator) Upgrade(ctx context.Context, in *proto.UpgradeRequest) (da
 		return data, err
 	}
 
+	// stop kubelet
+	if _, err = r.Stop(ctx, &proto.StopRequest{Id: "kubelet"}); err != nil {
+		return data, err
+	}
+
+	// kubeadm Reset
+	if err = upgrade.Reset(); err != nil {
+		return data, err
+	}
+
 	// Trigger reboot
+	// handle this towards the end of the stack so we can be sure
+	// we dont boot the node in an err situation
```

Comment:
If we don't want to reboot on an error then `defer` might not be what you want. This will always be executed, expect on a panic.